crates/wast/src/binary.rs

```diff
@@ -144,6 +153,110 @@ fn encode_fields(
     }
 }
 
+fn encode_component_fields(
+    component_id: &Option<Id<'_>>,
+    component_name: &Option<NameAnnotation<'_>>,
+    fields: &[ComponentField<'_>],
+) -> Vec<u8> {
+    use crate::ast::CustomPlace::*;
+    use crate::ast::CustomPlaceAnchor::*;
+
+    let mut types = Vec::new();
+    let mut imports = Vec::new();
+    let mut funcs = Vec::new();
+    let mut exports = Vec::new();
+    let mut start = Vec::new();
+    let mut customs = Vec::new();
+    let mut instances = Vec::new();
+    let mut modules = Vec::new();
+    let mut components = Vec::new();
+    let mut aliases = Vec::new();
+    for field in fields {
+        match field {
+            ComponentField::Type(i) => types.push(i),
+            ComponentField::Import(i) => imports.push(i),
+            ComponentField::Func(i) => funcs.push(i),
+            ComponentField::Export(i) => exports.push(i),
+            ComponentField::Start(i) => start.push(i),
+            ComponentField::Custom(i) => customs.push(i),
+            ComponentField::Instance(i) => instances.push(i),
+            ComponentField::NestedModule(i) => modules.push(i),
+            ComponentField::NestedComponent(i) => components.push(i),
+            ComponentField::Alias(a) => aliases.push(a),
+        }
+    }
+
+    let mut e = Encoder {
+        wasm: Vec::new(),
+        tmp: Vec::new(),
+        customs: &customs,
+    };
+    e.wasm.extend(b"\0asm");
+    e.wasm.extend(b"\x01\0\0\0");
+
+    e.custom_sections(BeforeFirst);
+
+    let mut items = fields
+        .iter()
+        .filter(|i| match i {
+            ComponentField::Alias(_)
+            | ComponentField::Type(_)
+            | ComponentField::Import(_)
+            | ComponentField::NestedComponent(_)
+            | ComponentField::Instance(_) => true,
+            _ => false,
+        })
+        .peekable();
+
+    // A special path is used for now to handle non-module-linking modules to
+    // work around WebAssembly/annotations#11
+    if aliases.len() == 0 && components.len() == 0 && instances.len() == 0 {
+        e.section_list(1, Type, &types);
+        e.section_list(2, Import, &imports);
+    } else {
+        while let Some(field) = items.next() {
```

Comment:
This is basically a weird affordance for module linking today, but I suspect that this will all go away for components. IIRC components have no internal orderings so I think that we'd create a component by simply iterating over the fields and emitting the items as we see them.